services/src/main/scala/cromwell/services/metadata/impl/archiver/ArchiveMetadataSchedulerActor.scala

```diff
@@ -0,0 +1,161 @@
+
+import org.apache.commons.csv.{CSVFormat, CSVPrinter}
+import akka.actor.{Actor, ActorLogging, ActorRef, Props}
+import akka.pattern.ask
+import akka.util.Timeout
+import common.util.StringUtil.EnhancedToStringable
+import common.util.TimeUtil.EnhancedOffsetDateTime
+import cromwell.core.{WorkflowAborted, WorkflowFailed, WorkflowId, WorkflowSucceeded}
+import cromwell.services.metadata.MetadataArchiveStatus.{Archived, Unarchived}
+import cromwell.services.metadata.MetadataService.{GetMetadataStreamAction, MetadataLookupStreamResponse, QueryForWorkflowsMatchingParameters, WorkflowQueryFailure, WorkflowQuerySuccess}
+import cromwell.services.metadata.WorkflowQueryKey._
+import cromwell.services.metadata.impl.archiver.ArchiveMetadataSchedulerActor._
+import cromwell.util.GracefulShutdownHelper
+import cromwell.util.GracefulShutdownHelper.ShutdownCommand
+import cromwell.database.sql.SqlConverters.{ClobOptionToRawString, TimestampToSystemOffsetDateTime}
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.concurrent.duration._
+import scala.util.{Failure, Success}
+import slick.basic.DatabasePublisher
+import cromwell.database.sql.tables.MetadataEntry
+import cromwell.services.metadata.MetadataQuery
+import java.io.OutputStreamWriter
+import java.nio.file.Files
+import java.nio.file.StandardOpenOption
+
+import cromwell.core.path.Path
+import cromwell.services.MetadataServicesStore
+import cromwell.services.metadata.impl.MetadataDatabaseAccess
+
+
+class ArchiveMetadataSchedulerActor(archiveMetadataConfig: ArchiveMetadataConfig,
+                                    serviceRegistryActor: ActorRef)
+  extends Actor
+    with ActorLogging
+    with GracefulShutdownHelper
+    with MetadataDatabaseAccess
+    with MetadataServicesStore {
+
+  implicit val ec: ExecutionContext = context.dispatcher
+  implicit val askTimeout: Timeout = new Timeout(60.seconds)
+
+  // kick off archiving immediately
+  self ! ArchiveNextWorkflowMessage
+
+  override def receive: Receive = {
+    case ArchiveNextWorkflowMessage => archiveNextWorkflow().onComplete({
+      case Success(true) => self ! ArchiveNextWorkflowMessage
+      case Success(false) => scheduleNextWorkflowToArchive()
+      case Failure(error) =>
+        log.error(error, s"Error while archiving, will retry.")
+        scheduleNextWorkflowToArchive()
+    })
+    case ShutdownCommand => context.stop(self)  // TODO: cancel any streaming that might be happening
+    case other => log.info(s"Programmer Error! The ArchiveMetadataSchedulerActor received unexpected message! ($sender sent $other})")
+  }
+
+  def archiveNextWorkflow(): Future[Boolean] = {
+    for {
+      maybeWorkflowId <- lookupNextWorkflowToArchive()
+      result <- maybeWorkflowId match {
+        case Some(id) => for {
+          path <- Future(archiveMetadataConfig.makePath(id))
+          _ = log.info(s"Archiving metadata for $id to ${path.pathAsString}")
+          dbStream <- fetchStreamFromDatabase(id)
```

Comment:
Do we need to close the stream in any situation (e.g. errors), or will it get cleaned up automatically when the reference goes out of scope?

Reply:
I looked into this, the stream doesn't actually have a `close` command.

My guess is the "stream" is actually just state in memory regarding which SQL command to send to the DB next, and therefore doesn't actually have any open resources outside of individual requests to the DB.